tests/quickcheck.rs

```diff
@@ -0,0 +1,326 @@
+extern crate num_bigint;
+extern crate num_integer;
+extern crate num_traits;
+
+#[cfg(test)]
+#[cfg(feature = "quickcheck")]
+#[macro_use]
+pub extern crate quickcheck;
+
+#[cfg(test)]
+#[cfg(feature = "quickcheck")]
+mod quickchecks {
+    use num_bigint::{BigInt, BigUint};
+    use num_traits::{Num, One, Pow, Zero};
+    use quickcheck::TestResult;
+
+    quickcheck! {
+        fn quickcheck_unsigned_eq_reflexive(a:BigUint) -> bool {
+            a == a
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_eq_reflexive(a:BigInt) -> bool {
+            a == a
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_eq_symmetric(a:BigUint, b:BigUint) -> bool {
+            if a == b {
+                b == a
+            } else {
+                b != a
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_eq_symmetric(a:BigInt, b:BigInt) -> bool {
+            if a == b {
+                b == a
+            } else {
+                b != a
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_add_primitive(a: u128, b: u128) -> TestResult {
+            let actual = BigUint::from(a) + BigUint::from(b);
+            match a.checked_add(b) {
+                None => TestResult::discard(),
+                Some(expected) => TestResult::from_bool(BigUint::from(expected) == actual)
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_signed_add_primitive(a: i128, b: i128) -> TestResult {
+            let actual = BigInt::from(a) + BigInt::from(b);
+            match a.checked_add(b) {
+                None => TestResult::discard(),
+                Some(expected) => TestResult::from_bool(BigInt::from(expected) == actual)
+            }
+        }
+    }
+
+    quickcheck! {
+        fn quickcheck_unsigned_add_commutative(a: BigUint, b: BigUint) -> bool {
+            a.clone() + b.clone() == b + a
```

Comment:
A lot of these `clone()` calls. could do their operation by reference instead, `&a + &b == b + a`, as long as you trust that ref/value operators are relating correctly. Or maybe that's good as another angle for quickcheck to poke at anyway.

Reply:
How much do you think I should trust that ref/value operators work properly? Isn't that implemented at a rust language level (i.e. there is no code in this crate to deal with it)? If that's the case, I don't feel the need to test it.

Reply:
It's not language level -- see `src/macros.rs` for what we use to implement value/reference combinations of operators. The forwarding is simple enough that I don't expect any bugs here, but you should also feel free to use it in testing (vs. clones).

Reply:
I think I'll use the reference operators. I expect it to test more code/branches. Two wrongs can't make a right 100 times in a row for random inputs. By which I mean that I don't have to trust both ref and value operators since ref operators are implemented in terms of value operators. If there is a bug in either the ref macros or the operator implementation I would probably catch it. Besides that, I expect the ref implementation to be faster, as it avoids needless copying.